cardano-api/test/Test/Cardano/Api/TxBody.hs

```diff
@@ -0,0 +1,233 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module Test.Cardano.Api.TxBody (tests) where
+
+import           Cardano.Prelude
+
+import           Hedgehog (Property, forAll, property, tripping)
+import           Test.Tasty (TestTree)
+import           Test.Tasty.Hedgehog (testProperty)
+import           Test.Tasty.TH (testGroupGenerator)
+
+import           Cardano.Api
+
+import           Test.Cardano.Api.Typed.Gen (genTxBody, genTxBodyContent)
+
+
+prop_roundtrip_TxBody_make_get_Byron :: Property
+prop_roundtrip_TxBody_make_get_Byron = roundtripTxBodyMakeGet ByronEra
+
+prop_roundtrip_TxBody_make_get_Shelley :: Property
+prop_roundtrip_TxBody_make_get_Shelley = roundtripTxBodyMakeGet ShelleyEra
+
+prop_roundtrip_TxBody_make_get_Allegra :: Property
+prop_roundtrip_TxBody_make_get_Allegra = roundtripTxBodyMakeGet AllegraEra
+
+prop_roundtrip_TxBody_make_get_Mary :: Property
+prop_roundtrip_TxBody_make_get_Mary = roundtripTxBodyMakeGet MaryEra
+
+-- TODO Alonzo
+-- prop_roundtrip_TxBody_make_get_Alonzo :: Property
+-- prop_roundtrip_TxBody_make_get_Alonzo = roundtripTxBodyMakeGet AlonzoEra
+
+
+roundtripTxBodyMakeGet :: IsCardanoEra era => CardanoEra era -> Property
+roundtripTxBodyMakeGet era =
+  property $ do
+    content <- forAll $ genTxBodyContent era
+    tripping
+      (normalizeOriginal $ review content)
+      (\_ -> makeTransactionBody content)
+      (<&> \(TxBody content') -> normalizeRoundtrip content')
+
+
+-- * normalize = strip unnecessary details
+
+normalizeOriginal :: TxBodyContent ViewTx era -> TxBodyContent ViewTx era
+normalizeOriginal content =
+  content
+    { txAuxScriptData = TxAuxScriptDataNone
+      -- TODO alonzo: this is just stub, remove it
+    , txAuxScripts    = normalizeAuxScripts   $ txAuxScripts   content
+    , txCertificates  = normalizeCertificates $ txCertificates content
+    , txIns           = sortOn fst            $ txIns          content
+    , txMetadata      = normalizeMetadata     $ txMetadata     content
+    , txMintValue     = normalizeMintValue    $ txMintValue    content
+    , txWithdrawals   = normalizeWithdrawals  $ txWithdrawals  content
+    }
+
+normalizeRoundtrip :: TxBodyContent ViewTx era -> TxBodyContent ViewTx era
+normalizeRoundtrip content@TxBodyContent{txIns} =
+  content
+    { txAuxScriptData = TxAuxScriptDataNone
+      -- TODO alonzo: this is just stub, remove it
+    , txIns           = sortOn fst txIns
+    }
+
+normalizeMetadata :: TxMetadataInEra era -> TxMetadataInEra era
+normalizeMetadata = \case
+  TxMetadataInEra _ (TxMetadata m) | null m -> TxMetadataNone
+  other                                     -> other
+
+normalizeAuxScripts :: TxAuxScripts era -> TxAuxScripts era
+normalizeAuxScripts = \case
+  TxAuxScripts _       []      -> TxAuxScriptsNone
+  TxAuxScripts support scripts ->
+    TxAuxScripts support $ map upgradeScriptInEra scripts
```

Comment:
Why are we upgrading the scripts?

Reply:
The instruction set from V1 may be used as V2, and we can't determine the language version from the transaction. When the user uses V1 instructions, we don't know if they used the V1 tag or the V2 tag. So it's safe to drop this information.